wonnx/templates/matrix/gemm.wgsl

```diff
@@ -1,71 +1,102 @@
-{%- include "structs.wgsl" -%}
+type Scalar = {{ scalar_type }};
+type GemmVec = vec{{ kernel_size }}<{{ scalar_type }}>;
+type GemmMat = mat{{ kernel_size }}x{{ kernel_size }}<{{ scalar_type }}>;
+
+struct GemmArrayVector {
+	data: array<GemmVec>;
+};
 
 [[group(0), binding(0)]]
-var<storage, read> input_0: ArrayVector;
+var<storage, read> input_0: GemmArrayVector;
 
 [[group(0), binding(1)]]
-var<storage, read> input_1: ArrayVector;
+var<storage, read> input_1: GemmArrayVector;
 
-{%- if i_lens | length == 3 -%} // Bias
-[[group(0), binding(2)]]
-var<storage, read> input_2: ArrayVector;
+{% if i_lens | length == 3 %} // Bias
+	[[group(0), binding(2)]]
+	var<storage, read> input_2: GemmArrayVector;
 
-[[group(0), binding(3)]]
-var<storage, write> output_0: ArrayVector;
-{%- else -%}
-[[group(0), binding(2)]]
-var<storage, write> output_0: ArrayVector;
-{%- endif -%}  
+	[[group(0), binding(3)]]
+	var<storage, write> output_0: GemmArrayVector;
+{% else %}
+	[[group(0), binding(2)]]
+	var<storage, write> output_0: GemmArrayVector;
+{% endif %}
 
 [[stage(compute), workgroup_size(1)]]
 fn main([[builtin(global_invocation_id)]] global_id: vec3<u32>) {
-	let y = global_id.x % {{ i_shape[1][1] / 4 | int }}u;
-	let x = global_id.x / {{ i_shape[1][1] / 4 | int }}u;
+	{% if i_shape[0][1] < kernel_size %}
+		let n_k = 1u;
```

Comment:
n could refer to too many things at this point. Having a more distinctive name like: `output_rows` for n, and explicating that you're doing a division might help. `output_rows_div_kernel`, is pretty verbose, but a lot easier to get than `n_k`